### ISSUE TITLE
Move ChainExtension trait to Contract common?

### ISSUE THREAD

**Issue by tylerztl** [paritytech/substrate#7846]
`ChainExtension` is an independent and public component, if moving it to contract common directory, it will reduce unnecessary dependencies.

**Comment by bkchr** [comment]
Cc @athei

**Comment by athei** [comment]
It is not independent. It depends on some types that are deeply intertwined with the rest of pallet contracts: `Ext, Error, Runtime, RuntimeToken`.

**Comment by atenjin** [comment]
> It is not independent. It depends on some types that are deeply intertwined with the rest of pallet contracts: `Ext, Error, Runtime, RuntimeToken`.

@athei 
Hi, we has this issue for we think the `ChainExtension` is a mechanism which is similar with the "Precompiled Contracts" in ETH, and the `function_id` is the id to call different things. Thus, we think we can allocate the `func_id` in a unified rule, then, developers who belong to different teams could develop their libs for different feature.
And then, the blockchain developer could combine different libs to implement their `ChainExtension` like:

```rust
fn call(func_id:u32, env: ...) {
    match func_id {
        0x01..0xff => {
            lib1::Lib1Extension::call(func_id, env)
        }
        0x01ff..0xffff => {
            lib2::Lib2Extension::call(func_id, env)
        }
         ...
    }
}
```
Thus the `lib1`, `lib2` need to impl `ChainExtension`, in this way, it could be like `unit` which could be combined in "composite pattern"( the lib and runtime ChainExtension are all implemented by `ChainExtension` trait).

So we advice that it maybe a good choose to move `ChainExtension` to `pallet-contracts-common` or some else common crate.

**Comment by athei** [comment]
>  is a mechanism which is similar with the "Precompiled Contracts" 

No. `ChainExtension` is not meant for composition with foreign code. It is to expose specifies if your own chain. Library contracts is something that is not yet implemented.

> So we advice that it maybe a good choose to move ChainExtension to pallet-contracts-common or some else common crate.

I don't see how this follows from the rest of your post.
